extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergTableAdapter.java

```diff
@@ -633,46 +655,312 @@ private static TableDefinition fromSchema(
+                relativePath = fileName;
+            }
+            final String newDataLocation = table.locationProvider().newDataLocation(relativePath);
+            ParquetTools.writeTable(dhTable, newDataLocation, parquetInstructions);
+        }
+        return parquetFilesWritten;
+    }
+
+    /**
+     * Commit the changes to the Iceberg table by creating a snapshot.
+     */
+    private void commit(
+            @NotNull final IcebergUtils.SpecAndSchema newSpecAndSchema,
+            @NotNull final Iterable<DataFile> appendFiles,
+            final boolean overwrite,
+            final boolean schemaVerified) {
+        final Transaction icebergTransaction = table.newTransaction();
+        final Snapshot currentSnapshot = table.currentSnapshot();
+        // For a null current snapshot, we are creating a new table. So we can just append instead of overwriting.
+        if (overwrite && currentSnapshot != null) {
+            // Fail if the table gets changed concurrently
+            final OverwriteFiles overwriteFiles = icebergTransaction.newOverwrite()
+                    .validateFromSnapshot(currentSnapshot.snapshotId())
+                    .validateNoConflictingDeletes()
+                    .validateNoConflictingData();
+
+            // Delete all the existing data files in the table
+            try (final Stream<DataFile> dataFiles = allDataFiles(table, currentSnapshot)) {
+                dataFiles.forEach(overwriteFiles::deleteFile);
+            }
+            appendFiles.forEach(overwriteFiles::addFile);
+            overwriteFiles.commit();
+
+            // Update the spec and schema of the existing table.
+            // If we have already verified the schema, we don't need to update it.
+            if (!schemaVerified) {
+                if (!table.schema().sameSchema(newSpecAndSchema.schema())) {
+                    final UpdateSchema updateSchema = icebergTransaction.updateSchema().allowIncompatibleChanges();
+                    table.schema().columns().stream()
+                            .map(Types.NestedField::name)
+                            .forEach(updateSchema::deleteColumn);
+                    newSpecAndSchema.schema().columns()
+                            .forEach(column -> updateSchema.addColumn(column.name(), column.type()));
+                    updateSchema.commit();
+                }
+                if (!table.spec().compatibleWith(newSpecAndSchema.partitionSpec())) {
+                    final UpdatePartitionSpec updateSpec = icebergTransaction.updateSpec();
+                    table.spec().fields().forEach(field -> updateSpec.removeField(field.name()));
+                    newSpecAndSchema.partitionSpec().fields().forEach(field -> updateSpec.addField(field.name()));
+                    updateSpec.commit();
+                }
+            }
+        } else {
+            // Append the new data files to the table
+            final AppendFiles append = icebergTransaction.newAppend();
```

Comment:
I don't think overwrite is special w.r.t. schema evolution. Append should probably have the same support.